executors.md

```diff
@@ -0,0 +1,179 @@
+# Summary
+[summary]: #summary
+
+This RFC proposes a design for `futures-executors`, including both executor
+traits and built-in executors. In addition, it sets up a core expectation for all
+tasks that they are able to spawn additional tasks, while giving fine-grained
+control over what executor that spawning is routed to.
+
+NOTE: this RFC assumes that [RFC #2] is accepted.
+
+[RFC #2]: https://github.com/rust-lang-nursery/futures-rfcs/pull/2
+
+# Motivation
+[motivation]: #motivation
+
+This is a follow-up to the [Tokio Reform], [Futures 0.2], and [Task Context]
+RFCs, harmonizing their designs.
+
+[Tokio Reform]: https://github.com/tokio-rs/tokio-rfcs/pull/3
+[Futures 0.2]: https://github.com/rust-lang-nursery/futures-rfcs/pull/1
+[Task Context]: https://github.com/rust-lang-nursery/futures-rfcs/pull/2
+
+The design in this RFC has the following goals:
+
+- Add a core assumption that tasks are *always* able to spawn additional tasks,
+  avoiding the need to reflect this at the API level. (Note: this assumption is
+  only made in contexts that also assume an allocator.)
+
+- Provide fine-grained control over which executor is used to fulfill that assumption.
+
+- Provide reasonable built-in executors for both single-threaded and
+  multithreaded execution.
+
+It brings the futures library into line with the design principles of the [Tokio
+Reform], but also makes some improvements to the `current_thread` design.
+
+# Proposed design
+
+## The core executor abstraction: `Spawn`
+
+First, in `futures-core` we provide a basic definition of a general-purpose
+executor:
+
+```rust
+trait Spawn {
+    fn spawn(&self, f: Box<Future<Item = (), Error = ()> + Send>);
+}
+```
+
+The `Spawn` trait is quite straightforward (some alternatives and tradeoffs are
+discussed in the next section), and crucially is object-safe.
+
+We then build in an executor to the task context, stored internally as a trait
+object, which allows us to provide the following methods:
+
+```rust
+impl task::Context {
+    // Spawn onto the current default executor
+    fn spawn<F>(&self, F)
+        where F: Future<Item = (), Error = ()> + Send + 'static;
+
+    // Run the given closure with a context using a new default executor
+    fn with_spawn<S, F, R>(&mut self, spawn: S, f: F) -> R
```

Comment:
Could you explain why `with_spawn` is a fn directly on context?

My initial thought on the matter is that you would want to set the spawn at context creation time (perhaps a builder) in the executor.

If one wants to swap a spawn in the midst of a task being polled, would it be possible to do that with a free function? This would add a bit of API pepper and also remove noise from the `Context` API docs.

I also haven't seen in the RFCs how [`with_notify`](https://docs.rs/futures/0.1.18/futures/executor/fn.with_notify.html) will be handled. I would guess it would be similar to whatever strategy `with_spawn` goes with.

Reply:
I see why tasks and executors are decoupled, but I'm still not sure why spawning is something a task does.

I know it kinda seems silly but I feel the same way about this situation as I do about regex libraries which have match, search, etc as methods on a regex object, unlike C++, which has match, search, etc as free functions.

And I'm not sure why the comment says that it's using a default executor... isn't spawn the executor? The comment would imply that there are now two executors involved somehow, spawn and the default one.

It's also not obvious why the spawn function takes a future and with_spawn takes a function (which will inevitably call poll with the context on a captured future).

Reply:
I had envisioned `with_notify` and `with_spawn` working the same way. I personally think it makes sense to have them directly on `Context` since they provide core functionality for modifying `Context`, and having them available as methods makes them easily discoverable.

> If one wants to swap a spawn in the midst of a task being polled, would it be possible to do that with a free function?

Sorry, can you give an example of what you mean? It should be possible to do something like this:

```rust
fn poll(...) -> ... {
    try_ready!(self.first_future.poll(ctx));
    ctx.with_spawn(self.my_spawn, |ctx| self.second_future.poll(ctx))
}
```

Is that what you had in mind?

Reply:
I'm slowly beginning to understand the overall design.

@cramertj You didn't address why the functions have different parameters, both of which are isomorphic (given the existence of `poll_fn`, and the lambda you wrote above).

Reply:
@ahmedcharles Sorry, I'm not sure I understand-- I'd expect that `with_spawn` and `with_notify` would have the same parameters, e.g.:

```rust
fn poll(...) -> ... {
    try_ready!(self.first_future.poll(ctx));
    ctx.with_notify(self.my_custom_notify, |ctx| self.second_future.poll(ctx));
    ctx.with_spawn(self.my_custom_spawn, |ctx| self.third_future.poll(ctx))
}
```

Reply:
`Context::spawn` and `Context::with_spawn` have different parameters. I wasn't comparing it to `with_notify`.

Reply:
@ahmedcharles To clarify-- `spawn` spawns a future onto the existing `Spawn`, while `with_spawn` runs a closure inside a new context with a new `Spawn`.